### PR TITLE
Added volume.Deleter interface and simple HostPath implementation

### PR DESCRIPTION
**Pull request by markturansky**
Part of https://github.com/kubernetes/kubernetes/pull/6773

volume.Deleter interface allows a DeletableVolumePlugin to remove volumes from the underlying infrastructure by setting pv.Spec.PersistentVolumeReclaimPolicy = "Delete"

This PR is the framework for Deleter and a deliberately gimped HostPath implementation for testing/example.  Each plugin would implement Deleter, as appropriate, in a separate PR.

@thockin @saad-ali @kubernetes/rh-storage

### CODE REVIEW

File: pkg/volume/host_path/host_path.go [hunk 1]
```diff
@@ -210,3 +233,24 @@ func (r *hostPathRecycler) Recycle() error {
...
+	return r.path
+}
+
+// Delete for hostPath removes the local directory so long as it is beneath /tmp.
+// Plugins with storage APIs (GCE, AWS, etc) should use the API for deletion of the resource from the infrastructure.
+func (r *hostPathDeleter) Delete() error {
+	if !strings.HasPrefix(r.GetPath(), "/tmp") {
+		return fmt.Errorf("host_path deleter only supports /tmp/*")
+	}
+	return os.RemoveAll(r.GetPath())
```
Comment (thockin) [inline pkg/volume/host_path/host_path.go]: This means a user can nuke all of /tmp - making me nervous here....

Comment (markturansky) [inline pkg/volume/host_path/host_path.go]: Will make it /tmp/hostpath or something else you suggest.

Comment (thockin) [inline pkg/volume/host_path/host_path.go]: Maybe just regex it as `/tmp/.+`  (or test that there is some element beyond /tmp/)

You know my feeling about using HostPath as a test vehicle for PV - the more powerful it becomes the more likely it is to cause some user to nuke their system.  Can we bolster the scary warning in the comment in PersistentVolumeSource?  The scarier the better. :)

**Comment by thockin** [comment]
LGTM but comment nits and path validation.

File: pkg/volume/host_path/host_path.go (hunk 1 above)
Comment (markturansky) [inline pkg/volume/host_path/host_path.go]: Scary comment included, fixed the path validation, and improved the validation tests.

**Comment by markturansky** [comment]
@thockin all green ... !

**Comment by markturansky** [comment]
@k8s-bot test this

File: pkg/volume/host_path/host_path.go (hunk 1 above)
Comment (thockin) [inline pkg/volume/host_path/host_path.go]: I was looking for a comment in the pkg/api/v1/types.go in PersistentVolumeSource, describing HostPath in even scarier terms.

**Comment by thockin** [comment]
I'm going to be out a lot of tomorrow, so assigned to saad for final review pass

File: pkg/volume/host_path/host_path.go (hunk 1 above)
Comment (markturansky) [inline pkg/volume/host_path/host_path.go]: I misread where you wanted the scary comment.  I'll add it to types.go.

**Comment by markturansky** [comment]
@saad-ali  I added a scarier warning message to internal/v1 types.go and in persistent-volumes.md doc.

All places where HostPath is used in PV contains "testing only" warnings, and now we've got "WILL NOT WORK" in a few places.  I believe the user is amply warned.

thanks for the review @thockin

**Comment by markturansky** [comment]
@saad-ali All green for this one.  Tim's feedback is implemented.

**Comment by saad-ali** [comment]
LGTM Thanks Mark.

Looks like you'll need to do another rebase though

**Comment by markturansky** [comment]
Rebased, fixed, and pushed anew.  Awaiting green.

**Comment by saad-ali** [comment]
@k8s-bot test this please

**Comment by saad-ali** [comment]
@k8s-bot test this please

**Comment by markturansky** [comment]
@saad-ali thanks!

**Pull request merged**